services/worker/src/worker/job_runners/split/duckdb_index.py

```diff
@@ -0,0 +1,280 @@
+        "size": repo_file.size,
+    }
+
+
+def compute_index_rows(
+    dataset: str,
+    config: str,
+    split: str,
+    duckdb_index_directory: StrPath,
+    target_revision: str,
+    hf_endpoint: str,
+    commit_message: str,
+    url_template: str,
+    hf_token: Optional[str],
+    committer_hf_token: Optional[str],
+) -> IndexRowsResponse:
+    logging.info(f"get split-duckdb-index for dataset={dataset} config={config} split={split}")
+
+    # validate split
+    split_names_best_response = get_previous_step_or_raise(
+        kinds=["config-split-names-from-streaming", "config-split-names-from-info"], dataset=dataset, config=config
+    )
+    try:
+        splits_content = split_names_best_response.response["content"]["splits"]
+    except Exception as e:
+        raise PreviousStepFormatError("Previous step did not return the expected content.", e) from e
+
+    if split not in [split_item["split"] for split_item in splits_content]:
+        raise SplitNotFoundError(f"The split '{split}' does not exist for the config '{config}' of the dataset.")
+
+    # get parquet content
+    config_parquet_best_response = get_previous_step_or_raise(kinds=["config-parquet"], dataset=dataset, config=config)
+    try:
+        parquet_files_content = config_parquet_best_response.response["content"]["parquet_files"]
+        parquet_file_items: List[ParquetFileItem] = [
+            parquet_file_item
+            for parquet_file_item in parquet_files_content
+            if parquet_file_item["config"] == config and parquet_file_item["split"] == split
+        ]
+        if not parquet_file_items:
+            raise ParquetResponseEmptyError("No parquet files found.")
+    except Exception as e:
+        raise PreviousStepFormatError("Previous step did not return the expected content.") from e
+
+    fs = HTTPFileSystem()
+    parquet_urls = [parquet_file_item["url"] for parquet_file_item in parquet_file_items]
+    desc = f"{dataset}/{config}"
+    try:
+        parquet_files: List[ParquetFile] = thread_map(
+            partial(get_parquet_file, fs=fs, hf_token=hf_token), parquet_urls, desc=desc, unit="pq", disable=True
+        )
+    except Exception as e:
+        raise FileSystemError(f"Could not read the parquet files: {e}") from e
+
+    # get the features
+    features = Features.from_arrow_schema(parquet_files[0].schema.to_arrow_schema())
+
+    # look for string columns using the first rows
+    string_columns = [column for column, feature in features.items() if STRING_FEATURE_DTYPE in str(feature)]
+
+    if not string_columns:
+        raise NoIndexableColumnsError("No string columns available to index.")
+
+    # look for image, audio and binary columns, if present, raise exeception (not supported yet)
+    if any(
+        feature
+        for feature in features.values()
+        if next(
+            (feature_type for feature_type in UNSUPPORTED_FEATURES_MAGIC_STRINGS if feature_type in str(feature)), None
```

Comment:
It does not break indexing but I was thinking that it will need preparation like we do in first rows when mapping image/audio to a file in assets or maybe something similar like in /rows (It surely will support algo pagination)

Reply:
But maybe I could leave indexing also image/binary/audio and then when implementing API layer think about it or apply any change needed, what do you think @severo ?

Reply:
Sorry, I don't understand. As any other non-text column, they will be ignored by this step, so... why do they receive a special treatment?

Reply:
Maybe it is something I am overthinking but my idea is that when using the index on the API layer, it will require something like https://github.com/huggingface/datasets-server/blob/main/services/api/src/api/routes/rows.py#L144 or https://github.com/huggingface/datasets-server/blob/main/services/worker/src/worker/job_runners/split/first_rows_from_parquet.py#L37 or https://github.com/huggingface/datasets-server/blob/main/services/worker/src/worker/job_runners/split/first_rows_from_streaming.py#L42 right?
I think when implementing /search  endpoint, it will download the index (cache for a certain time period), search the specified text and return the dataframe result of it (no need to read from parquet, just from duckdb index file), I wonder how will we return those datasets with image/audio/binary data to the viewer?

Reply:
> no need to read from parquet, just from duckdb index file

is this how the index works? I thought that it would only contain an index for the text columns, while having to access the parquet file to get the full rows (the other columns, basically). Isn't it how it works?

Reply:
No, like a typical database engine, it needs to have all the data in a table and then it adds an index to it.

Reply:
hmmm... OK. And how will we manage sharded parquet files? will the index be sharded too? Or is it a huge file? is it what we want?

Reply:
At least we create different tables/index files (one per parquet file) and, in API layer, join all the results 
what do you think @huggingface/datasets-server  @lhoestq ?

Reply:
I think one big duckdb file on the first 5GB is fine for now. The size is available in the `config-parquet` response afaik

Reply:
Is it easy to truncate by size?

Most of the datasets should have parquet files under 5GB, but we have:
- false positives for datasets with scripts, where we don't detect the size is above 5GB (like https://huggingface.co/datasets/Muennighoff/multi_eurlex)
- parquet-native datasets (like https://huggingface.co/datasets/bigcode/the-stack) that can be enormous

For those, we would have to cut somewhere.

Reply:
For now, I limited to those splits with parquets under 5 GB (sum of all shards). I think  it can scale later to cut to the first N GB?

Reply:
It is going to be limited to datasets with parquets under 100 MB initially (small), then we can think of a better way to scale for bigger datasets.